### PR TITLE
feat: add `upstash-redis` driver

### PR DESCRIPTION
**Pull request by Hebilicious**
### 🔗 Linked issue

Resolves https://github.com/unjs/unstorage/issues/288

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation, readme, or JSdoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

This PR adds a upstash-redis driver for redis  support on edge platforms. 

For full context, when vercel-kv was added in https://github.com/unjs/unstorage/pull/213, @redis/upstash was added as a dependency because it is a vercel-kv dependency, so that makes it easier for pnpm users.

This PR adds a redis-upstash driver for the use case where you want to use upstash without vercel (which I encountered recently)

image

The tests are running locally, and need env variables to work in the CI. 
We should create an unjs upstash db, or mock the client. 

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by chronark** [comment]
Hey there!
(disclaimer: I work for upstash)

quick question: why did you install `@vercel/kv` and then use `@upstash/redis`
It seems to work cause their package is just a wrapper around `@upstash/redis` but it seemed weird :D

secondly: we can provde a db for unit tests for free, just let me know :)

**Comment by Hebilicious** [comment]
> Hey there! (disclaimer: I work for upstash)
> 
> quick question: why did you install `@vercel/kv` and then use `@upstash/redis` It seems to work cause their package is just a wrapper around `@upstash/redis` but it seemed weird :D
> 
> secondly: we can provde a db for unit tests for free, just let me know :)

Hello there! @vercel/kv is already a driver supported by unstorage, this PR doesn't install it. I didn't know their package was just a wrapper.

On providing us with a DB, I think that would be great ! 
@pi0 is the maintainer of unstorage, so he'll need to green light this

**Comment by Atinux** [comment]
Do you mind resolving the conflicts @Hebilicious ?

**Comment by Hebilicious** [comment]
> Do you mind resolving the conflicts @Hebilicious ?

Sure. For full context, when vercel-kv was added in https://github.com/unjs/unstorage/pull/213, @redis/upstash was added as a dependency because it is a vercel-kv dependency, so that makes it easier for pnpm users. 

This PR adds a redis-upstash driver for the use case where you want to use upstash without vercel (which I encountered recently)

I'll add this to the PR description.

File: .env.example
Comment (pi0) [inline .env.example:2]: please remove these / any local testing bits

File: docs/content/6.drivers/redis.md
Comment (pi0) [inline docs/content/6.drivers/redis.md:62]: This section shall be in `upstash.md`

File: package.json
Comment (pi0) [inline package.json:79]: remove this

File: src/drivers/redis-upstash.ts
Comment (pi0) [inline src/drivers/redis-upstash.ts:1]: We can simplify driver name to upstash with an alias to upstash-redis as this is only upstash/kv storage option now.

File: src/drivers/redis-upstash.ts
Comment (pi0) [inline src/drivers/redis-upstash.ts:4]: Interface should be `UpstashRedisOptions`

File: src/types.ts
Comment (pi0) [inline src/types.ts:37]: This is out of scope of this PR

**Review by pi0 — CHANGES_REQUESTED** [review]

**Comment by pi0** [comment]
Hi dear @chronark thanks for reaching out and for your offer about the CI plan.

I would ideally love to have a proper "unit" test for both drivers. I think we can mock the HTTP client with upstash API. If you have a tool or package that can help in this regard it would be awesome! Also, feel free to reach me via discord (pi0) if you would like to work closer on upstash integrations I can invite you to the unjs discord channel.

---

Off topic: Regarding vercel-kv, the main reason for having a wrapper is that it allows having "zero configuration" integration with their platform / provided environment variables - also any changes needed in the future. If there are other hosting options that provided upstash integration, we probably have to either use their packages or make a wrapper driver.

**Comment by pi0** [comment]
Thanks for the updates @Hebilicious ❤️

**Comment by chronark** [comment]
@pi0 we don't have any true unit test library or similar, I'm running all of our tests against a real upstash db or in a docker container :/

I'm leaving upstash soon, joining your discord now probably won't help much unfortunately.

if you want credits for testing against real dbs, just let us know here: support@upstash.com and ideally reference this comment :D

**Comment by pi0** [comment]
Landed via https://github.com/unjs/unstorage/pull/500

**Pull request closed without merging**
